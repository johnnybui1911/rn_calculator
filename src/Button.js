import React from 'react'
import { Text, TouchableOpacity } from 'react-native'
import { connect } from 'react-redux'
import { pressNumber, pressClear, pressEqual, pressOperator, pressDot } from './actions'
import styles from './styles'

const Button = ({ item, isOperator, pressNumber, pressEqual, pressClear, pressOperator, pressDot }) => {
  const handlePress = () => {
    if (typeof item === 'number') {
      pressNumber(item)
    } else {
      switch (item) {
        case '=': {
          pressEqual()
          break
        }
        case 'C': {
          pressClear()
          break
        }
        case '.': {
          pressDot()
          break
        }
        default: {
          pressOperator(item)
          break
        }
      }
    }
  }

  return (
    <TouchableOpacity key={item} style={[styles.button, { marginLeft: isOperator ? 6 : 0 }]} onPress={handlePress}>
      <Text style={styles.input}>{item}</Text>
    </TouchableOpacity>
  )
}

const MemoButton = React.memo(Button)
const mapDispatchToProps = {
  pressNumber,
  pressClear,
  pressEqual,
  pressOperator,
  pressDot
}
export default connect(null, mapDispatchToProps)(MemoButton)
