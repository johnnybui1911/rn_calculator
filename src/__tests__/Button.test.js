import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Button } from '../Button'

describe('Button', () => {
  describe('Button Rendering', () => {
    const mockData = {
      item: 1,
      isOperator: true
    }
    let wrapper = shallow(<Button {...mockData} />)
    it('should match to snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })

  let [pressNumber, pressEqual, pressClear, pressOperator, pressDot] = new Array(5).fill(jest.fn())
  describe('Button Interaction Dispatch Action', () => {
    let props

    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('should call the press number function', () => {
      props = {
        item: 1,
        isOperator: true,
        pressNumber
      }
      wrapper = shallow(<Button {...props} />)
      const render = wrapper.find('TouchableOpacity').first()
      render.props().onPress()
      expect(pressNumber).toHaveBeenCalled()
      expect(pressNumber).toHaveBeenCalledTimes(1)
      expect(pressNumber).toHaveBeenCalledWith(1)
    })

    it('should call the press operator function', () => {
      const operator = '+'
      props = {
        item: operator,
        isOperator: false,
        pressOperator
      }
      wrapper = shallow(<Button {...props} />)
      const render = wrapper.find('TouchableOpacity').first()
      render.props().onPress()
      expect(pressOperator).toHaveBeenCalled()
      expect(pressOperator).toHaveBeenCalledTimes(1)
      expect(pressOperator).toHaveBeenCalledWith(operator)
    })

    it('should call the press equal function', () => {
      props = {
        item: '=',
        isOperator: false,
        pressEqual
      }
      wrapper = shallow(<Button {...props} />)
      const render = wrapper.find('TouchableOpacity').first()
      render.props().onPress()
      expect(pressEqual).toHaveBeenCalled()
      expect(pressEqual).toHaveBeenCalledTimes(1)
    })

    it('should call the press clear function', () => {
      props = {
        item: 'C',
        isOperator: false,
        pressClear
      }
      wrapper = shallow(<Button {...props} />)
      const render = wrapper.find('TouchableOpacity').first()
      render.props().onPress()
      expect(pressClear).toHaveBeenCalled()
      expect(pressClear).toHaveBeenCalledTimes(1)
    })

    it('should call the press dot function', () => {
      props = {
        item: '.',
        isOperator: false,
        pressDot
      }
      wrapper = shallow(<Button {...props} />)
      const render = wrapper.find('TouchableOpacity').first()
      render.props().onPress()
      expect(pressDot).toHaveBeenCalled()
      expect(pressDot).toHaveBeenCalledTimes(1)
    })
  })
})
