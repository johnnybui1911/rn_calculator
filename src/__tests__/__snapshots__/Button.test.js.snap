// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button Rendering should match to snapshot 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  key="1"
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#d9d9d9",
        "borderRadius": 6,
        "flex": 1,
        "justifyContent": "center",
        "margin": 6,
      },
      Object {
        "marginLeft": 0,
      },
    ]
  }
>
  <Text
    style={
      Object {
        "fontSize": 24,
      }
    }
  >
    1
  </Text>
</TouchableOpacity>
`;
